File: libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/navigation/RouteTileDownloaderTest.java
package com.mapbox.services.android.navigation.v5.navigation;

import org.junit.Test;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class RouteTileDownloaderTest {

  @Test
  public void startDownload_fetchRouteTilesIsCalled() {
    String tilePath = "some/path/";
    OfflineNavigator offlineNavigator = mock(OfflineNavigator.class);
    RouteTileDownloadListener listener = mock(RouteTileDownloadListener.class);
    OfflineTiles offlineTiles = mock(OfflineTiles.class);
    when(offlineTiles.version()).thenReturn("some-version");
    RouteTileDownloader downloader = new RouteTileDownloader(offlineNavigator, tilePath, listener);

    downloader.startDownload(offlineTiles);

    verify(offlineTiles).fetchRouteTiles(any(TarFetchedCallback.class));
  }

  @Test
  public void onError_downloadListenerErrorTriggered() {
    String tilePath = "some/path/";
    OfflineNavigator offlineNavigator = mock(OfflineNavigator.class);
    RouteTileDownloadListener listener = mock(RouteTileDownloadListener.class);
    OfflineError offlineError = mock(OfflineError.class);
    RouteTileDownloader downloader = new RouteTileDownloader(offlineNavigator, tilePath, listener);

    downloader.onError(offlineError);

    verify(listener).onError(offlineError);
  }
}

File: libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/navigation/RouteTileDownloaderTest.kt
package com.mapbox.services.android.navigation.v5.navigation

import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.junit.Test

class RouteTileDownloaderTest {

    @Test
    fun startDownload_fetchRouteTilesIsCalled() {
        val tilePath = "some/path/"
        val offlineTiles = mockk<OfflineTiles>(relaxed = true)
        every { offlineTiles.version() } returns "some-version"
        val downloader = RouteTileDownloader(mockk(), tilePath, mockk())

        downloader.startDownload(offlineTiles)

        verify { offlineTiles.fetchRouteTiles(any()) }
    }

    @Test
    fun onError_downloadListenerErrorTriggered() {
        val tilePath = "some/path/"
        val listener = mockk<RouteTileDownloadListener>(relaxed = true)
        val offlineError = mockk<OfflineError>()
        val downloader = RouteTileDownloader(mockk(), tilePath, listener)

        downloader.onError(offlineError)

        verify { listener.onError(offlineError) }
    }
}
